### ISSUE TITLE
Support returning structs from functions

### ISSUE THREAD

**Issue by skywinder** [BANKEX/web3swift#33]
Version 0.4.22 still gets issues with that;

https://github.com/ethereum/solidity/issues/40
FYI:
https://ethereum.stackexchange.com/questions/7317/how-can-i-return-struct-when-function-is-called
https://github.com/ethereum/solidity/issues/1601
https://github.com/ethereum/solidity/issues/2948
https://github.com/ethereum/solidity/pull/2863

Depends on #32

**Comment by kyriediculous** [comment]
Is there an ETA for this , next protocol release / Solidity 0.5.0?

**Comment by shamatar** [comment]
EncoderV2 is in 0.4.0 already, you can specify to use it with the following code
```
let contract = web3.contract(jsonString, at: nil, abiVersion: 2)
```

Although I didn't have a chance to test it well in read world yet, requires some beta Solidity compiler voodoo. So far it's backward compatible with V1 anyway

**Comment by kyriediculous** [comment]
I've tried it with the 0.4.22 compiler.
It seems to not play nice with remix and web3 still? Last thing I read is that it only worked with Ether.js so far. Not sure if there have been any updates on this since 29th of jan.

**Comment by shamatar** [comment]
Hello @kyriediculous 

I've made the following contract 
```
pragma solidity ^0.4.21;

pragma experimental ABIEncoderV2;

contract Tester {
    struct TestStruct {
        uint256 number;
        string someText;
        uint256[2] staticArray;
        uint256[] dynamicArray;
        string[2] anotherDynamicArray;
    }
    
    
    constructor() {
        
    }
    
    function testSingle() public pure returns (TestStruct t) {
        uint256[] memory dynArray = new uint256[](3);
        dynArray[0] = 0;
        dynArray[1] = 1;
        dynArray[2] = 2;
        uint256[2] memory stArray = [uint256(0), uint256(1)];
        string[2] memory anotherDynArray = ["Hello", "World"];
        t = TestStruct({
            number: 1,
            someText: "Hello world",
            staticArray: stArray,
            dynamicArray: dynArray,
            anotherDynamicArray: anotherDynArray
        });
        return t;
    }
    
    function testDynArray() public pure returns (TestStruct[] ts) {
        ts = new TestStruct[](2);
        ts[0] = testSingle();
        ts[1] = testSingle();
    }
    
    function testStaticArray() public pure returns (TestStruct[2] ts) {
        ts[0] = testSingle();
        ts[1] = testSingle();
    }
}
```

So far the latest version from @develop branch successfully works with "testSingle" method.

Sincerely, Alex

**Comment by shamatar** [comment]
Full support should be in [master](/bankex/web3swift/) now

**Comment by shamatar** [comment]
Full encoding support is also in [master](/bankex/web3swift/) now

**Comment by kyriediculous** [comment]
It's likely that this is compatible with web3 1.0.0 right now but not with anything else; Not with remix or web3 implementations that are compatible with Metamask (since they are web3 1.0.0 incompatible) as of writing. 

Of course this isn't at your end.

Remix: 
```
[call] from:0xca35b7d915458ef540ade6068dfe2f44e8fa733c, to:Tester.testSingle(), data:f376e...6e013, return: 
{
	"error": "Failed to decode output: Error: Unsupported or invalid type: tuple"
}
```

**Comment by shamatar** [comment]
Yes, as far as I know remix and web3 are based on ethereumjs-abi JavaScript package and do not support decoding yet (etherjs does support V2 encoding). Nevertheless you can follow an ABI spec from solidity documents directly and implement it. I’ve published this contract and was calling a function as a part of the test, Ethereum network itself works with contracts with experimental ABI, and 0.4.22 looks quite stable with it (they had to implement usage of “returndatasize” assembler instructions to make it possible to return a structure, that was merged as a part of 0.4.22), so I expect ABIv2 to become stable and non-experimental somewhere in 0.4.23 or 0.5 that should come quite soon.

You can look at the tests for web3swift and find a full ABI and contract address on Rinkeby. Any pull requests to add more exhaustive tests and edge cases are welcome.

Sincerely, Alex

________________________________

**Comment by roschler** [comment]
If I deploy a contract using the ABI experimental function(s) described here for returning a struct from a Soldity function, will it work with Web3.js on Rinkeby and the main net?

**Comment by shamatar** [comment]
Hello @roschler

First of all, ABIv2 encoder/decoder in web3swift is backward compatible with a normal one. It will understand all current encoding types, but will also decode experimental features, such as returning structs. You can already use these features if you want by putting this #pragma is your code, and it will be included in 0.5 version of solidity. Interestingly, solidity docs describing ABI encoding imply features of ABIv2, looks like those features just were not implemented yet in main solidity branch.

So, if you do not need this features - just remove a pragma and everything will work. If you do need them, you will have to supply a pragma and use V2 decoder from web3swift.

Sincerely, Alex
